src/pdf.sandbox.js

```diff
@@ -24,66 +25,67 @@ const TESTING =
   typeof PDFJSDev === "undefined" || PDFJSDev.test("!PRODUCTION || TESTING");
 
 class Sandbox {
-  constructor(module) {
+  constructor(module, extraNames) {
     this._evalInSandbox = module.cwrap("evalInSandbox", null, [
       "string",
       "int",
     ]);
-    this._dispatchEventName = null;
+    this._commFun = module.cwrap("commFun", null, ["string", "string", "int"]);
     this._module = module;
-    this._alertOnError = 1;
+    this._alertOnError = 0;
+    this._extraNames = extraNames;
   }
 
   create(data) {
     const sandboxData = JSON.stringify(data);
-    const extra = [
-      "send",
-      "setTimeout",
-      "clearTimeout",
-      "setInterval",
-      "clearInterval",
-      "crackURL",
-    ];
+    const extra = ["send", "globalEval"].concat(this._extraNames);
     const extraStr = extra.join(",");
-    let code = [
-      "exports = Object.create(null);",
-      "module = Object.create(null);",
+    const code = [
       // Next line is replaced by code from initialization.js
       // when we create the bundle for the sandbox.
       PDFJSDev.eval("PDF_SCRIPTING_JS_SOURCE"),
-      `data = ${sandboxData};`,
-      `module.exports.initSandbox({ data, extra: {${extraStr}}, out: this});`,
-      "delete exports;",
-      "delete module;",
-      "delete data;",
+      `pdfjsScripting.initSandbox({ data: ${sandboxData}, extra: {${extraStr}}, out: this})`,
```

Comment:
Is your intention of this to hide these extra helpers from the script? That is not accomplished by this implementation. Your code basically becomes:

```
pdfjsScripting.initSandbox({ // ...
  extra: {
    send, globalEval, ...
  },
})
```
Above syntax is "shorthand property names" - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#New_notations_in_ECMAScript_2015
In ES5 syntax, it is:
```
pdfjsScripting.initSandbox({ // ...
  extra: {
    send: send,
    globalEval: globalEval,
    // ...
  },
})
```

Now it should be more obvious that this basically reading off the global (since `send`, `globalEval`, etc. aren't default global variables.)

Reply:
These functions are global variables in the sandbox, for example:
https://github.com/mozilla/pdf.js.quickjs/blob/895a07683ae7dc71484ce59f493a5c2e52900a7b/src/quickjs_extra.c#L61-L63
These functions are wrappers to call their counterpart in pdf.sandbox.external.js.